# 1. Reading NumPy arrays

In [1]:
import numpy as np
x = np.genfromtxt("ionosphere.txt", delimiter=",")
x[:3]

array([[ 1.     ,  0.     ,  0.99539, -0.05889,  0.85243,  0.02306,
         0.83398, -0.37708,  1.     ,  0.0376 ,  0.85243, -0.17755,
         0.59755, -0.44945,  0.60536, -0.38223,  0.84356, -0.38542,
         0.58212, -0.32192,  0.56971, -0.29674,  0.36946, -0.47357,
         0.56811, -0.51171,  0.41078, -0.46168,  0.21266, -0.3409 ,
         0.42267, -0.54487,  0.18641, -0.453  ,  1.     ],
       [ 1.     ,  0.     ,  1.     , -0.18829,  0.93035, -0.36156,
        -0.10868, -0.93597,  1.     , -0.04549,  0.50874, -0.67743,
         0.34432, -0.69707, -0.51685, -0.97515,  0.05499, -0.62237,
         0.33109, -1.     , -0.13151, -0.453  , -0.18056, -0.35734,
        -0.20332, -0.26569, -0.20468, -0.18401, -0.1904 , -0.11593,
        -0.16626, -0.06288, -0.13738, -0.02447, -1.     ],
       [ 1.     ,  0.     ,  1.     , -0.03365,  1.     ,  0.00485,
         1.     , -0.12062,  0.88965,  0.01198,  0.73082,  0.05346,
         0.85443,  0.00827,  0.54591,  0.00299,  0.83775, -0.13644

In [2]:
x = x = np.genfromtxt("ionosphere.txt", delimiter=",", usecols=np.arange(34))
x[:3]

array([[ 1.     ,  0.     ,  0.99539, -0.05889,  0.85243,  0.02306,
         0.83398, -0.37708,  1.     ,  0.0376 ,  0.85243, -0.17755,
         0.59755, -0.44945,  0.60536, -0.38223,  0.84356, -0.38542,
         0.58212, -0.32192,  0.56971, -0.29674,  0.36946, -0.47357,
         0.56811, -0.51171,  0.41078, -0.46168,  0.21266, -0.3409 ,
         0.42267, -0.54487,  0.18641, -0.453  ],
       [ 1.     ,  0.     ,  1.     , -0.18829,  0.93035, -0.36156,
        -0.10868, -0.93597,  1.     , -0.04549,  0.50874, -0.67743,
         0.34432, -0.69707, -0.51685, -0.97515,  0.05499, -0.62237,
         0.33109, -1.     , -0.13151, -0.453  , -0.18056, -0.35734,
        -0.20332, -0.26569, -0.20468, -0.18401, -0.1904 , -0.11593,
        -0.16626, -0.06288, -0.13738, -0.02447],
       [ 1.     ,  0.     ,  1.     , -0.03365,  1.     ,  0.00485,
         1.     , -0.12062,  0.88965,  0.01198,  0.73082,  0.05346,
         0.85443,  0.00827,  0.54591,  0.00299,  0.83775, -0.13644,
         0.75535, 

In [3]:
y = np.genfromtxt("ionosphere.txt", delimiter=",", usecols=34, dtype='int')
y[:3]

array([ 1, -1,  1])

# 2. Functions in Python

In [4]:
def cube(x):
    """this is my first function in this course; it cubes a number"""
    return x**3
cube(3)

27

In [5]:
help(cube)

Help on function cube in module __main__:

cube(x)
    this is my first function in this course; it cubes a number



In [6]:
def powers(x):
    """this function computes both square and cube"""
    return x**2, x**3
powers(3)

(9, 27)

In [7]:
powers(3)[1]

27

# 3. Assignment in Python

In [8]:
a = np.array([1,2,3])
a

array([1, 2, 3])

In [9]:
b = a
b

array([1, 2, 3])

In [10]:
a[1] = 100
a

array([  1, 100,   3])

In [11]:
b

array([  1, 100,   3])

In [12]:
b = np.array(a)

In [13]:
b = a.copy()

# 4. Rank of a number

In [14]:
a = np.array([1,2,3])
b = np.array([5,6])
np.concatenate((a,b))

array([1, 2, 3, 5, 6])

In [15]:
my_array = np.concatenate((np.arange(10**6),np.array([4])))

In [16]:
print(my_array)

[     0      1      2 ... 999998 999999      4]


Exercise 1: Describe briefly the array my_array in English.

Answer: 1 Million numbers from 0 to 999,999 followed by 4

Exercise 2: Why do we have pairs of closing and opening parentheses after "concatenate"?

Answer: we have an extra pair of parentheses so we can add other arguments. For example concatenate((a,b), axis=0, out=None) can be done without confusing it with concatenation elements.

In [17]:
import time
start = time.time()
print(sum(4 >= my_array))
print(time.time() - start,"seconds")

6
1.9686479568481445 seconds


Exercise 3. Why do you think the result above is 6?

Answer: Because the first five elements (0,1,2,3,4) and the last element (4) are <= 4.

In [19]:
start = time.time()
count = 0
for n in range(my_array.size):
    if 4 >= my_array[n]:
        count = count + 1
print(count)
print(time.time() - start,"seconds")

6
0.33307313919067383 seconds


# 5. Computing Minimum

In [21]:
import math
start = time.time()
current_minimum = math.inf
for n in range(my_array.size):
    if current_minimum > my_array[n]:
        current_minimum = my_array[n]
print(current_minimum)
print(time.time() - start,"seconds")

0
0.2430572509765625 seconds


Exercise 4. Modify  the  code  to  also  find  the  position  of  the  minimum  (i.e. the index of the smallest element of the array; you may assume that there is aunique smallest element).

In [22]:
current_minimum = math.inf
its_position = math.inf
for n in range(my_array.size):
    if current_minimum > my_array[n]:
        current_minimum = my_array[n]
        its_position = n
print(current_minimum, its_position)

0 0


# 7. Other exercises

In [29]:
#5.

def euclidean_norm(array,length):
    """Euclidean length"""
    v = 0
    for i in range(length):
        v = v+array[i]**2
    return math.sqrt(v)
euclidean_norm([1,2],2)

2.23606797749979

In [36]:
#6.

def euclidean_norm(array):
    """Euclidean length"""
    v = 0
    for i in range(array.size):
        v = v+array[i]**2
    return math.sqrt(v)
z = np.array([1,2])
euclidean_norm(z)

2.23606797749979